java.lang.IllegalAccessError: Failed to unreflect constructor

Hi! I downloaded an OpenJ9 build to try it out. Adopt OpenJDK doesn't has a "release" for the OpenJDK 10 + OpenJ9 combination, so I downloaded the Linux x86_64 nightly of 28 June 2018 from here:

https://adoptopenjdk.net/nightly.html?variant=openjdk10-openj9

I have a hobby project that uses Nashorn in some places, but it was failing with an exception on OpenJ9. On the other hand, this code works fine on OpenJDK 10 10.0.1+10-4 from the Debian testing amd64 repositories.

Test case:

```java
System.setProperty( "nashorn.debug", "true" );
final String[] options = { "--language=es6", "-strict" };
var factory = new NashornScriptEngineFactory();
var engine = (NashornScriptEngine) factory.getScriptEngine( options );

try {
	engine.put( "javaList", Arrays.asList( new Object[] {1,2,3} ) );
	engine.eval( "javaList.forEach(function(i){print(i);})" );
} catch ( ScriptException e ) {
	throw new RuntimeException( e );
}
```
Exception thrown in this test case:

```
Exception in thread "main" java.lang.IllegalAccessError: Failed to unreflect constructor public jdk.nashorn.javaadapters.java_lang_Iterable(jdk.nashorn.internal.runtime.ScriptObject)
	at jdk.dynalink.linker.support.Lookup.unreflectConstructor(jdk.dynalink@10.0.1/Lookup.java:245)
	at jdk.dynalink.beans.CallerSensitiveDynamicMethod.unreflectConstructor(jdk.dynalink@10.0.1/CallerSensitiveDynamicMethod.java:185)
	at jdk.dynalink.beans.CallerSensitiveDynamicMethod.getTarget(jdk.dynalink@10.0.1/CallerSensitiveDynamicMethod.java:171)
	at jdk.dynalink.beans.OverloadedDynamicMethod.getInvocation(jdk.dynalink@10.0.1/OverloadedDynamicMethod.java:220)
	at jdk.dynalink.beans.StaticClassLinker$SingleClassStaticsLinker.getGuardedInvocation(jdk.dynalink@10.0.1/StaticClassLinker.java:157)
	at jdk.dynalink.beans.StaticClassLinker.getGuardedInvocation(jdk.dynalink@10.0.1/StaticClassLinker.java:207)
	at jdk.dynalink.beans.BeansLinker$NoSuchMemberHandlerBindingLinker.getGuardedInvocation(jdk.dynalink@10.0.1/BeansLinker.java:226)
	at jdk.nashorn.internal.runtime.linker.NashornBeansLinker.getGuardedInvocation(jdk.scripting.nashorn@10.0.1/NashornBeansLinker.java:169)
	at jdk.nashorn.internal.runtime.linker.NashornStaticClassLinker.delegate(jdk.scripting.nashorn@10.0.1/NashornStaticClassLinker.java:111)
	at jdk.nashorn.internal.runtime.linker.NashornStaticClassLinker.getGuardedInvocation(jdk.scripting.nashorn@10.0.1/NashornStaticClassLinker.java:104)
	at jdk.dynalink.linker.support.CompositeTypeBasedGuardingDynamicLinker.getGuardedInvocation(jdk.dynalink@10.0.1/CompositeTypeBasedGuardingDynamicLinker.java:184)
	at jdk.dynalink.linker.support.CompositeGuardingDynamicLinker.getGuardedInvocation(jdk.dynalink@10.0.1/CompositeGuardingDynamicLinker.java:132)
	at jdk.dynalink.LinkerServicesImpl.lambda$getGuardedInvocation$0(jdk.dynalink@10.0.1/LinkerServicesImpl.java:160)
	at jdk.dynalink.LinkerServicesImpl$$Lambda$67.000000004849E550.get(Unknown Source)
	at jdk.dynalink.LinkerServicesImpl.getWithLookupInternal(jdk.dynalink@10.0.1/LinkerServicesImpl.java:191)
	at jdk.dynalink.LinkerServicesImpl.getGuardedInvocation(jdk.dynalink@10.0.1/LinkerServicesImpl.java:158)
	at jdk.nashorn.internal.runtime.linker.JavaAdapterFactory.getConstructor(jdk.scripting.nashorn@10.0.1/JavaAdapterFactory.java:194)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getSamTypeConverter(jdk.scripting.nashorn@10.0.1/NashornLinker.java:168)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.convertToType(jdk.scripting.nashorn@10.0.1/NashornLinker.java:118)
	at jdk.dynalink.TypeConverterFactory.createConverter(jdk.dynalink@10.0.1/TypeConverterFactory.java:405)
	at jdk.dynalink.TypeConverterFactory$1$1.computeValue(jdk.dynalink@10.0.1/TypeConverterFactory.java:126)
	at jdk.dynalink.TypeConverterFactory$1$1.computeValue(jdk.dynalink@10.0.1/TypeConverterFactory.java:122)
	at jdk.dynalink.ClassMap.get(jdk.dynalink@10.0.1/ClassMap.java:155)
	at jdk.dynalink.TypeConverterFactory.getCacheableTypeConverter(jdk.dynalink@10.0.1/TypeConverterFactory.java:360)
	at jdk.dynalink.TypeConverterFactory.getCacheableTypeConverterNull(jdk.dynalink@10.0.1/TypeConverterFactory.java:347)
	at jdk.dynalink.TypeConverterFactory.getTypeConverterNull(jdk.dynalink@10.0.1/TypeConverterFactory.java:353)
	at jdk.dynalink.TypeConverterFactory.asType(jdk.dynalink@10.0.1/TypeConverterFactory.java:250)
	at jdk.dynalink.LinkerServicesImpl.asType(jdk.dynalink@10.0.1/LinkerServicesImpl.java:131)
	at jdk.nashorn.internal.runtime.linker.NashornBeansLinker$NashornBeansLinkerServices.asType(jdk.scripting.nashorn@10.0.1/NashornBeansLinker.java:259)
	at jdk.dynalink.beans.LinkerServicesWithMissingMemberHandlerFactory.asType(jdk.dynalink@10.0.1/LinkerServicesWithMissingMemberHandlerFactory.java:55)
	at jdk.dynalink.beans.LinkerServicesWithMissingMemberHandlerFactory.asType(jdk.dynalink@10.0.1/LinkerServicesWithMissingMemberHandlerFactory.java:55)
	at jdk.dynalink.linker.LinkerServices.asTypeLosslessReturn(jdk.dynalink@10.0.1/LinkerServices.java:145)
	at jdk.dynalink.beans.SingleDynamicMethod.createConvertingInvocation(jdk.dynalink@10.0.1/SingleDynamicMethod.java:295)
	at jdk.dynalink.beans.SingleDynamicMethod.getInvocation(jdk.dynalink@10.0.1/SingleDynamicMethod.java:195)
	at jdk.dynalink.beans.SingleDynamicMethod.lambda$getInvocation$0(jdk.dynalink@10.0.1/SingleDynamicMethod.java:131)
	at jdk.dynalink.beans.SingleDynamicMethod$$Lambda$91.000000004877A5B0.get(Unknown Source)
	at jdk.dynalink.LinkerServicesImpl.getWithLookupInternal(jdk.dynalink@10.0.1/LinkerServicesImpl.java:191)
	at jdk.dynalink.LinkerServicesImpl.getWithLookup(jdk.dynalink@10.0.1/LinkerServicesImpl.java:179)
	at jdk.nashorn.internal.runtime.linker.NashornBeansLinker$NashornBeansLinkerServices.getWithLookup(jdk.scripting.nashorn@10.0.1/NashornBeansLinker.java:298)
	at jdk.dynalink.beans.LinkerServicesWithMissingMemberHandlerFactory.getWithLookup(jdk.dynalink@10.0.1/LinkerServicesWithMissingMemberHandlerFactory.java:85)
	at jdk.dynalink.beans.LinkerServicesWithMissingMemberHandlerFactory.getWithLookup(jdk.dynalink@10.0.1/LinkerServicesWithMissingMemberHandlerFactory.java:85)
	at jdk.dynalink.beans.SingleDynamicMethod.getInvocation(jdk.dynalink@10.0.1/SingleDynamicMethod.java:131)
	at jdk.dynalink.beans.DynamicMethodLinker.getGuardedInvocation(jdk.dynalink@10.0.1/DynamicMethodLinker.java:124)
	at jdk.dynalink.beans.BeansLinker$NoSuchMemberHandlerBindingLinker.getGuardedInvocation(jdk.dynalink@10.0.1/BeansLinker.java:226)
	at jdk.dynalink.beans.BeansLinker.getGuardedInvocation(jdk.dynalink@10.0.1/BeansLinker.java:356)
	at jdk.nashorn.internal.runtime.linker.NashornBeansLinker.getGuardedInvocation(jdk.scripting.nashorn@10.0.1/NashornBeansLinker.java:169)
	at jdk.nashorn.internal.runtime.linker.NashornBeansLinker.getGuardedInvocation(jdk.scripting.nashorn@10.0.1/NashornBeansLinker.java:156)
	at jdk.dynalink.linker.support.CompositeGuardingDynamicLinker.getGuardedInvocation(jdk.dynalink@10.0.1/CompositeGuardingDynamicLinker.java:132)
	at jdk.dynalink.LinkerServicesImpl.lambda$getGuardedInvocation$0(jdk.dynalink@10.0.1/LinkerServicesImpl.java:160)
	at jdk.dynalink.LinkerServicesImpl$$Lambda$67.000000004849E550.get(Unknown Source)
	at jdk.dynalink.LinkerServicesImpl.getWithLookupInternal(jdk.dynalink@10.0.1/LinkerServicesImpl.java:191)
	at jdk.dynalink.LinkerServicesImpl.getGuardedInvocation(jdk.dynalink@10.0.1/LinkerServicesImpl.java:158)
	at jdk.dynalink.DynamicLinker.relink(jdk.dynalink@10.0.1/DynamicLinker.java:265)
	at jdk.nashorn.internal.scripts.Script$Recompilation$1$\^eval\_.0000000048236380.:program(<eval>:1)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(jdk.scripting.nashorn@10.0.1/ScriptFunctionData.java:655)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(jdk.scripting.nashorn@10.0.1/ScriptFunction.java:513)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(jdk.scripting.nashorn@10.0.1/ScriptRuntime.java:527)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(jdk.scripting.nashorn@10.0.1/NashornScriptEngine.java:448)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(jdk.scripting.nashorn@10.0.1/NashornScriptEngine.java:405)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(jdk.scripting.nashorn@10.0.1/NashornScriptEngine.java:401)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(jdk.scripting.nashorn@10.0.1/NashornScriptEngine.java:154)
	at javax.script.AbstractScriptEngine.eval(java.scripting@10.0.1/AbstractScriptEngine.java:264)
	at test.Testing.main(Testing.java:424)
Caused by: java.lang.IllegalAccessException: 'jdk.scripting.nashorn.scripts' no access to: 'jdk.nashorn.javaadapters'
	at java.lang.invoke.MethodHandles$Lookup.checkClassModuleVisibility(java.base@10.0.1/MethodHandles.java:660)
	at java.lang.invoke.MethodHandles$Lookup.checkClassAccess(java.base@10.0.1/MethodHandles.java:448)
	at java.lang.invoke.MethodHandles$Lookup.checkAccess(java.base@10.0.1/MethodHandles.java:358)
	at java.lang.invoke.MethodHandles$Lookup.checkAccess(java.base@10.0.1/MethodHandles.java:314)
	at java.lang.invoke.MethodHandles$Lookup.unreflectConstructor(java.base@10.0.1/MethodHandles.java:1144)
	at jdk.dynalink.linker.support.Lookup.unreflectConstructor(jdk.dynalink@10.0.1/Lookup.java:243)
	... 62 more
```

--- a/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleInfoImpl.java
+++ b/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleInfoImpl.java
@@ -93,7 +93,7 @@ final class MethodHandleInfoImpl implements MethodHandleInfo {
 			throw new NullPointerException();
 		}
 		try {
-			lookup.checkAccess(mh);
+			lookup.checkAccess(mh, false);
 		} catch (IllegalAccessException e) {
 			/*[MSG "K0583", "The Member is not accessible to the Lookup object"]*/
 			IllegalArgumentException x = new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0583")); //$NON-NLS-1$

--- a/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java
+++ b/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java
@@ -257,7 +257,7 @@ public class MethodHandles {
 				handle = handle.asFixedArity(); // remove unnecessary varargsCollector from the middle of the MH chain.
 				handle = convertToVarargsIfRequired(handle.bindTo(receiver));
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), receiverClass, handle.getModifiers());
 			
 			handle = SecurityFrameInjector.wrapHandleWithInjectedSecurityFrameIfRequired(this, handle);
@@ -337,7 +337,7 @@ public class MethodHandles {
 			return false;
 		}
 		
-		void checkAccess(MethodHandle handle) throws IllegalAccessException {
+		void checkAccess(MethodHandle handle, boolean skipAccessCheckPara) throws IllegalAccessException {
 			if (INTERNAL_PRIVILEGED == accessMode) {
 				// Full access for use by MH implementation.
 				return;
@@ -347,11 +347,11 @@ public class MethodHandles {
 				throw new IllegalAccessException(this.toString());
 			}
 
-			checkAccess(handle.getDefc(), handle.getMethodName(), handle.getModifiers(), handle);
+			checkAccess(handle.getDefc(), handle.getMethodName(), handle.getModifiers(), handle, skipAccessCheckPara);
 		}
 		
 		/*[IF Sidecar19-SE]*/
-		void checkAccess(VarHandle handle) throws IllegalAccessException {
+		void checkAccess(VarHandle handle, boolean reflectiveAccess) throws IllegalAccessException {
 			if (INTERNAL_PRIVILEGED == accessMode) {
 				// Full access for use by MH implementation.
 				return;
@@ -361,7 +361,7 @@ public class MethodHandles {
 				throw new IllegalAccessException(this.toString());
 			}
 			
-			checkAccess(handle.getDefiningClass(), handle.getFieldName(), handle.getModifiers(), null);
+			checkAccess(handle.getDefiningClass(), handle.getFieldName(), handle.getModifiers(), null, reflectiveAccess);
 		}
 		
 		void checkAccess(Class<?> clazz) throws IllegalAccessException {
@@ -389,14 +389,15 @@ public class MethodHandles {
 		 * 				 The object is always {@link MethodHandle} in the current implementation, so in order 
 		 * 				 to avoid extra type checking, the parameter is {@link MethodHandle}. If we need to
 		 * 				 handle {@link VarHandle} in the future, the type can be {@link Object}.
+		 * @param skipAccessCheckPara skip access check for MethodType parameters
 		 * @throws IllegalAccessException If the member is not accessible.
 		 * @throws IllegalAccessError If a handle argument or return type is not accessible.
 		 */
-		private void checkAccess(Class<?> targetClass, String name, int memberModifiers, MethodHandle handle) throws IllegalAccessException {
+		private void checkAccess(Class<?> targetClass, String name, int memberModifiers, MethodHandle handle, boolean skipAccessCheckPara) throws IllegalAccessException {
 			checkClassAccess(targetClass);
 
 			/*[IF Sidecar19-SE]*/
-			if (null != handle) {
+			if (null != handle && !skipAccessCheckPara) {
 				MethodType type = handle.type();
 				Module accessModule = accessClass.getModule();
 
@@ -580,7 +581,7 @@ public class MethodHandles {
 					/*[MSG "K0586", "Lookup class ({0}) must be the same as or subclass of the current class ({1})"]*/
 					throw new IllegalAccessException(com.ibm.oti.util.Msg.getString("K0586", accessClass, handleDefc)); //$NON-NLS-1$
 				}
-				checkAccess(handle);
+				checkAccess(handle, false);
 				checkSecurity(handleDefc, clazz, handle.getModifiers());
 				handle = SecurityFrameInjector.wrapHandleWithInjectedSecurityFrameIfRequired(this, handle);
 			} catch (DefaultMethodConflictException e) {
@@ -641,7 +642,7 @@ public class MethodHandles {
 				handle = new DirectHandle(clazz, methodName, type, MethodHandle.KIND_STATIC, null);
 				handle = HandleCache.putMethodInPerClassCache(cache, methodName, type, convertToVarargsIfRequired(handle));
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), clazz, handle.getModifiers());
 			/* Static check is performed by native code */
 			handle = SecurityFrameInjector.wrapHandleWithInjectedSecurityFrameIfRequired(this, handle);
@@ -717,7 +718,7 @@ public class MethodHandles {
 				HandleCache.putMethodInPerClassCache(cache, methodName, type, handle);
 			}
 			handle = restrictReceiver(handle);
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), clazz, handle.getModifiers());
 			handle = SecurityFrameInjector.wrapHandleWithInjectedSecurityFrameIfRequired(this, handle);
 			return handle;
@@ -924,7 +925,7 @@ public class MethodHandles {
 				handle = new FieldGetterHandle(clazz, fieldName, fieldType, accessClass);
 				HandleCache.putFieldInPerClassCache(cache, fieldName, fieldType, handle);
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), clazz, handle.getModifiers());
 			return handle;
 		}
@@ -951,7 +952,7 @@ public class MethodHandles {
 				handle = new StaticFieldGetterHandle(clazz, fieldName, fieldType, accessClass);
 				HandleCache.putFieldInPerClassCache(cache, fieldName, fieldType, handle);
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), clazz, handle.getModifiers());
 			return handle;
 		}
@@ -986,7 +987,7 @@ public class MethodHandles {
 				/*[MSG "K05cf", "illegal setter on final field"]*/
 				throw new IllegalAccessException(Msg.getString("K05cf")); //$NON-NLS-1$
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), clazz, handle.getModifiers());
 			return handle;
 		}
@@ -1021,7 +1022,7 @@ public class MethodHandles {
 				/*[MSG "K05cf", "illegal setter on final field"]*/
 				throw new IllegalAccessException(Msg.getString("K05cf")); //$NON-NLS-1$
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), clazz, handle.getModifiers());
 			return handle;
 		}
@@ -1235,7 +1236,7 @@ public class MethodHandles {
 			
 			if (!method.isAccessible()) {
 				handle = restrictReceiver(handle);
-				checkAccess(handle);
+				checkAccess(handle, true);
 			}
 			
 			handle = SecurityFrameInjector.wrapHandleWithInjectedSecurityFrameIfRequired(this, handle);
@@ -1319,7 +1320,7 @@ public class MethodHandles {
 			} 
 			
 			if (!method.isAccessible()) {
-				checkAccess(handle);
+				checkAccess(handle, true);
 			}
 			
 			return handle;
@@ -1350,7 +1351,7 @@ public class MethodHandles {
 				}
 				handle = HandleCache.putMethodInPerClassCache(cache, "<init>", type, convertToVarargsIfRequired(handle)); //$NON-NLS-1$
 			}
-			checkAccess(handle);
+			checkAccess(handle, false);
 			checkSecurity(handle.getDefc(), declaringClass, handle.getModifiers());
 			return handle;
 		}
@@ -1382,7 +1383,7 @@ public class MethodHandles {
 			}
 			
 			if (!method.isAccessible()) {
-				checkAccess(handle);
+				checkAccess(handle, true);
 			}
 			
 			handle = SecurityFrameInjector.wrapHandleWithInjectedSecurityFrameIfRequired(this, handle);
@@ -1427,7 +1428,7 @@ public class MethodHandles {
 				HandleCache.putFieldInPerClassCache(cache, fieldName, fieldType, handle);
 			}			
 			if (!field.isAccessible()) {
-				checkAccess(handle);
+				checkAccess(handle, true);
 			}
 			return handle;
 		}
@@ -1478,7 +1479,7 @@ public class MethodHandles {
 			}
 			
 			if (!field.isAccessible()) {
-				checkAccess(handle);
+				checkAccess(handle, true);
 			}
 			return handle;
 		}
@@ -1503,7 +1504,7 @@ public class MethodHandles {
 				}
 			}
 			try {
-				checkAccess(target);
+				checkAccess(target, false);
 				checkSecurity(target.getDefc(), target.getReferenceClass(), target.getModifiers());
 			} catch (IllegalAccessException e) {
 				throw new IllegalArgumentException(e);
@@ -1674,7 +1675,7 @@ public class MethodHandles {
 				} else {
 					handle = new InstanceFieldVarHandle(definingClass, name, type, this.accessClass);
 				}
-				checkAccess(handle);
+				checkAccess(handle, false);
 				checkSecurity(definingClass, definingClass, handle.getModifiers());
 				return handle;
 			} catch (NoSuchFieldError e) {
@@ -1710,7 +1711,7 @@ public class MethodHandles {
 				handle = new InstanceFieldVarHandle(field, field.getDeclaringClass(), field.getType());
 			}
 			
-			checkAccess(handle);
+			checkAccess(handle, true);
 			checkSecurity(handle.getDefiningClass(), handle.getDefiningClass(), handle.modifiers);
 			
 			return handle;

